File: 2015/day7/day7.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <unordered_map>
#include <vector>
#include <typeinfo>
using namespace std;

unordered_map<string, vector<string>> parseInput(const string &filename)
{
    ifstream file(filename);
    unordered_map<string, vector<string>> instructions;
    string line;

    while (getline(file, line))
    {
        stringstream ss(line);
        string token, lhs, op, rhs, arrow, dest;
        vector<string> parts;

        ss >> lhs;
        if (lhs == "NOT")
        {
            // "NOT x -> y"
            ss >> rhs >> arrow >> dest;
            parts = {"NOT", rhs}; // Store operation and operand
        }
        else
        {
            ss >> op;
            if (op == "->")
            {
                // "123 -> x" or "a -> b"
                dest = op; // Assign dest first
                parts = {"ASSIGN", lhs};
            }
            else
            {
                // "x AND y -> z"
                ss >> rhs >> arrow >> dest;
                parts = {op, lhs, rhs}; // Store full operation
            }
        }

        instructions[dest] = parts; // Store parsed instruction
    }

    return instructions;
}

int main()
{

    string filename = "input.txt";
    unordered_map<string, vector<string>> instructions = parseInput(filename);
    unordered_map<string, int> memo = {};
    // Debugging: Print parsed instructions
    for (const auto &[wire, expr] : instructions)
    {
        cout << wire << " ";
        for (const auto &part : expr)
            cout << part << " ";
        cout << endl;
    }

    return 0;
}

File: 2015/day7/day7.js
const fs = require("fs");

const filename = "input.txt";
const instructions = parseInput(filename);
let memo = new Map();

function parseInput(filename) {
    const instructions = new Map();
    const lines = fs.readFileSync(filename, "utf-8").trim().split("\n");

    for (const line of lines) {
        const parts = line.split(" ");
        let dest, expr;

        if (parts[0] === "NOT") {
            // "NOT x -> y"
            dest = parts[3];
            expr = ["NOT", parts[1]];
        } else if (parts[1] === "->") {
            // "123 -> x" or "a -> b"
            dest = parts[2];
            expr = ["ASSIGN", parts[0]];
        } else {
            // "x AND y -> z"
            dest = parts[4];
            expr = [parts[1], parts[0], parts[2]];
        }

        instructions.set(dest, expr);
    }

    return instructions;
}

function get_value(token,){
    
    let value = /^\d+$/.test(token) ? parseInt(token, 10) : token;

    if(typeof(value) == 'number'){
        return value
    }
    
    if (memo.has(token)) {
        return memo.get(token);
    }

    let data = instructions.get(token)

    if (!data) {
        console.error(`No instruction found for token: ${token}`);
        return 0;
    }

    let val;
    
    if(data[0] == 'ASSIGN'){
        val = get_value(data[1])
    } else if (data[0] === 'NOT') {
        val = ~get_value(data[1]) & 0xFFFF
    }else{
        let op = data[0]
        let rightValue = get_value(data[2])
        let leftValue = get_value(data[1])
        if(op == "AND"){
            val = (leftValue & rightValue ) & 0xFFFF
        }else if(op == "OR"){
            val = (leftValue | rightValue ) & 0xFFFF
        }else if(op == "LSHIFT"){
            val = (leftValue << rightValue ) & 0xFFFF
        }else if(op == "RSHIFT"){
            val = (leftValue >> rightValue ) & 0xFFFF
        }
    }
    memo.set(token, val);
    return val
}
console.log(instructions.get('b'))
let b = get_value('a')
instructions.set('b', ['ASSIGN', b.toString()])
console.log(instructions.get('b'))
memo.clear()
console.log(get_value('a'))






